casper-deploy-notifier/src/lib.rs

```diff
@@ -0,0 +1,97 @@
+use casper_types::AsymmetricType;
+use eventsource_stream::{Event, EventStreamError, Eventsource};
+use futures::stream::{BoxStream, TryStreamExt};
+use tokio::sync::mpsc;
+
+use crate::error::SseError;
+use crate::sse_types::SseData;
+use crate::types::Notification;
+
+mod error;
+mod sse_types;
+mod types;
+
+const DEFAULT_SSE_SERVER: &str = "https://events.mainnet.casperlabs.io";
+const DEFAULT_EVENT_CHANNEL: &str = "/events/main";
+
+type BoxedEventStream = BoxStream<'static, Result<Event, EventStreamError<reqwest::Error>>>;
+
+pub struct DeployNotifier {
+    url: String,
+}
+
+impl Default for DeployNotifier {
+    fn default() -> Self {
+        let url = format!("{}{}", DEFAULT_SSE_SERVER, DEFAULT_EVENT_CHANNEL);
+        Self { url }
+    }
+}
+
+impl DeployNotifier {
+    pub fn new(url: &str) -> Self {
+        DeployNotifier {
+            url: url.to_string(),
+        }
+    }
+
+    async fn connect(&self) -> Result<BoxedEventStream, SseError> {
+        // Connect to SSE endpoint.
+        let client = reqwest::Client::new();
+        let response = client.get(&self.url).send().await?;
+
+        let stream = response.bytes_stream();
+        let mut event_stream = stream.eventsource();
+
+        // Handle the handshake with API version.
+        let handshake_event = event_stream
+            .try_next()
+            .await?
+            .ok_or(SseError::StreamExhausted)?;
+        let handshake_data: SseData = serde_json::from_str(&handshake_event.data)?;
+        let _api_version = match handshake_data {
+            SseData::ApiVersion(v) => Ok(v),
+            _ => Err(SseError::InvalidHandshake),
+        }?;
+
+        // Wrap stream with box.
+        let boxed_event_stream = Box::pin(event_stream);
+
+        Ok(boxed_event_stream)
+    }
+
+    // Handle incoming events - look for successfuly processed deployments.
+    pub async fn run(&mut self, tx: mpsc::Sender<Notification>) -> Result<(), SseError> {
```

Comment:
I would rename it to `subscribe`

Reply:
Maybe I am biased about `subscribe`, but just take a look at API of Casper SDKs:

```
watcher.subscribe(deploy_a);
watcher.subscribe(deploy_b);
watcher.subscribe(deploy_c);
let result = watcher.start().await; // runs until short timeout is reached, or event about deploy_a/b/c is received.
```

Therefore I think `run` is better for this long running function.

Reply:
I don't think we should strive to follow the Casper SDK design decisions, as they seem to result in imperative and non-composable code.
The current API
```rust
let notifier = DeployNotifier { url, event_stream: None };
notifier.connect(); // mutates the DeployNotifier to now be conneced.
notifier.run(notification_sender);
```
Suggestion:
```rust
subscribe("http://casper-node.com", notification_sender);
```